### PR TITLE
Metadata system lazor

### PR DESCRIPTION
**Pull request by Julien00859**

### CODE REVIEW

File: incipyt/__main__.py [hunk 1]
```diff
@@ -76,41 +98,45 @@ def choice_callback(_ctx, _param, _choice):
...
-        metadata_setter.default("PROJECT_NAME", folder.resolve().name)
+        project.environ.suggest("PROJECT_NAME", folder.resolve().name)
     else:
         if (folder.is_absolute() and folder.is_dir() and any(folder.iterdir())) or (
             ("." / folder).is_dir() and any(("." / folder).resolve().iterdir())
         ):
             raise click.BadArgumentUsage(f"FOLDER {folder} is not empty.")
-        metadata_setter.default("PROJECT_NAME", folder.name)
-
-    project.environ["LICENSE"] = license
```
Comment (Julien00859) [inline incipyt/__main__.py]: the license is now part of the `**kwargs` and injected inside the environment cli source via `feed_environ`

File: incipyt/__main__.py [hunk 2]
```diff
@@ -122,6 +148,41 @@ def main(folder, verbose, silent, vcs, env, build, license, **kwargs):  # noqa:
     logger.info("All done.")
 
 
+def feed_environ(options, kwargs):
+    project.environ.feed_cli(dict([option.partition("=")[::2] for option in options]))
```
Comment (Julien00859) [inline incipyt/__main__.py:167]: The moment the "obvious" solution to 5AM me is `option.partition("=")[::2]` and not `option.split("=", 1)`...

File: incipyt/__main__.py [hunk 3]
```diff
@@ -165,11 +226,10 @@ def setup_logging(verbosity):
...
-# python -m incipyt <command>
-if sys.path[0] in ("", os.getcwd()):  # noqa: PTH109
-    sys.path.pop(0)
-
 if __name__ == "__main__":
+    # Remove '' and current working directory from the first entry of
+    # sys.path, if present to avoid using current directory in incipyt
+    # commands, when invoked as python -m incipyt <command>
+    if sys.path[0] in ("", os.getcwd()):  # noqa: PTH109
+        sys.path.pop(0)
```
Comment (Julien00859) [inline incipyt/__main__.py:252]: this should not be part of "feat: metadata" but in his own commit

File: incipyt/project/environment.py [hunk 4]
```diff
@@ -41,25 +75,45 @@ def __getitem__(self, key):
 
     def __setitem__(self, key, value):
         if key in self.data:
-            raise ValueError(f"Environment variable {key} already exists, delete it before.")
-
-        logger.debug("Set environment variable %s=%s.", key, value)
+            raise ValueError(f"Environment variable {key} is set already, delete it before.")
+        logger.debug("set %s=%s.", key, value)
```
Comment (Julien00859) [inline incipyt/project/environment.py:81]: Useless diff

File: incipyt/project/environment.py [hunk 5]
```diff
@@ -41,25 +75,45 @@ def __getitem__(self, key):
...
-            )
-        if user_input:
-            return user_input
-        else:
-            return (
-                user_input
-                if key in variables.metadata and variables.metadata[key].required
-                else None
-            )
+        if not (var := variables.get(key)):
```
Comment (Julien00859) [inline incipyt/project/environment.py:85]: Using warlus means you only intent to use the variable inside of the block, you use it outside, thus is should be defined outiside

File: incipyt/project/environment.py [hunk 6]
```diff
@@ -41,25 +75,45 @@ def __getitem__(self, key):
...
+        # No default value found, prompt user
+        click_prompt = partial(click.prompt, key.replace("_", " ").capitalize(), type=var.type)
+        if not var.required:
+            click_prompt.keywords["default"] = ""
+            return click_prompt() or None  # it asks only once
+        return click_prompt()  # it keeps asking until the user gives a value
+
+    def inject(self, key, value):
+        """Inject a confirmed key=value pair inside the environment."""
+        self._source_tool_confirmed[key] = value
```
Comment (Julien00859) [inline incipyt/project/environment.py:102]: Now there's two way of injecting a value, we should only keep one:
```py
environ[key] = value
```
and
```py
environ.inject(key, value)
```

the former takes precedence on the cli/osenv, the latter does not. But this is not obvious and I'm not sure it will be more useful than confusing

File: incipyt/project/meta_variables.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+    When there is no `default` suggestion, if no value is given when prompt asks for it
+    `None` is saved in `environ` and all :class:`incipyt._internal.templates.Formattable`
+    values will be purge from template dictionary before commiting. When `True`, an empty
+    string is saved in `environ`. Note that if the `default` suggestion is not empty,
+    `required` has no effect as the prompt result will never be an empty string.
+    """
+
+    name: str
+    default: Any = ""
```
Comment (Julien00859) [inline incipyt/project/meta_variables.py:60]: We should make the default read-only, so people update the environ, not the defaults

File: incipyt/project/meta_variables.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    prompt: bool = True
+    required: bool = False
```
Comment (Julien00859) [inline incipyt/project/meta_variables.py:62]: There is a weird relationship between prompt and required.. hope it doesn't get weirder in future versions

File: incipyt/project/meta_variables.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+logger = logging.getLogger(__name__)
+
+# the public index where every variable metadata is saved
+variables = {}
+
+
+@dataclasses.dataclass
+class Variable:
+    """Metadata dataclass for environment variables.
```
Comment (Julien00859) [inline incipyt/project/meta_variables.py:17]: there are missing bits from the documentation

File: incipyt/project/meta_variables.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+Variable("PACKAGE_VERSION", default="0.0.0")
+Variable("PROJECT_NAME")
+Variable("PYTHON_CMD", default=sys.executable, prompt=False)
+Variable("SUMMARY_DESCRIPTION", required=True)
+Variable("YEAR", type=int, default=date.today().year, prompt=False)
+
+Variable("CHECK_BUILD", type=strtobool, default=False, prompt=False)
+Variable("VENV_FOLDER", default=".venv", prompt=False)
+
+Variable("REPOSITORY")
```
Comment (Julien00859) [inline incipyt/project/meta_variables.py]: We should sort them alphabetically

File: incipyt/project/meta_variables.py [hunk 11]
```diff
@@ -60,6 +60,8 @@ class Variable:
     default: Any = ""
     prompt: bool = True
     required: bool = False
+    help: str = ""  # noqa: A003
```
Comment (Julien00859) [inline incipyt/project/meta_variables.py:63]: Missing documentation

File: incipyt/project/meta_variables.py [hunk 12]
```diff
@@ -60,6 +60,8 @@ class Variable:
...
+    export: bool = True
```
Comment (Julien00859) [inline incipyt/project/meta_variables.py:64]: Unused yet, the idea was that in case one to generate a basic configuration, the `export=False` keys shouldn't be exported to the configuration, but we only create an empty file for now so I guess that's for a future PR

File: incipyt/tools/git.py [hunk 13]
```diff
@@ -33,9 +33,10 @@ def add_to_structure(self):
             - Documentation: {REPOSITORY}/wiki
         """
         project.structure.use_template("python.gitignore", ".gitignore")
-        signals.project_url.emit(url_kind="Documentation", url_value="{REPOSITORY}/wiki")
-        signals.project_url.emit(url_kind="Issue", url_value="{REPOSITORY}/issues")
         signals.project_url.emit(url_kind="Repository", url_value="{REPOSITORY}")
+        if "github" in (project.environ["REPOSITORY"] or "").lower():
+            signals.project_url.emit(url_kind="Documentation", url_value="{REPOSITORY}/wiki")
+            signals.project_url.emit(url_kind="Issue", url_value="{REPOSITORY}/issues")
```
Comment (Julien00859) [inline incipyt/tools/git.py:39]: move to another PR

File: tests/test_renderers.py [hunk 14]
```diff
@@ -1,98 +1,166 @@
...
 from incipyt._internal.templates import FormatterEnviron
+from incipyt.project.meta_variables import Variable
 from tests.utils import mock_stdin
 
 
+@fixture(scope="function")
+def patch_variables():
+    try:
+        project.variables.clear()
+        variables = project.variables.copy()
```
Comment (Julien00859) [inline tests/test_renderers.py]: wrong order

File: incipyt/tools/venv.py [hunk 15]
```diff
@@ -1,22 +1,23 @@
...
+        signals.vcs_ignore.emit(pattern="{VENV_FOLDER}/")
 
     def pre(self, workon):
         """Run `python -m venv .env`.
 
         :param workon: Work-on folder.
         :type workon: :class:`pathlib.Path`
         """
-        env_path = workon / ".env"
+        env_path = workon / FormatterEnviron().format("{VENV_FOLDER}")
```
Comment (Julien00859) [inline incipyt/tools/venv.py:20]: I think this is a complicated way to write
```suggestion
        env_path = workon / project.environ["{VENV_FOLDER}"]
```

Don't forget to import project

File: incipyt/project/environment.py
Comment (fblanchetNaN) [inline incipyt/project/environment.py:1]: Maybe some comments (formal docstring or juste comments) on what are the different sources are missing.

File: incipyt/project/environment.py (hunk 6 above)
Comment (fblanchetNaN) [reply incipyt/project/environment.py:108]: At least it should be documented to limit the confusion.

File: incipyt/project/structure.py [hunk 16]
```diff
@@ -234,8 +101,8 @@ def mkdir(self, workon):
             logger.info("Mkdir folders for %s.", str(config_root))
             config_root.mkdir()
 
-    @staticmethod
-    def _visit(template):
+    @classmethod
+    def _visit(cls, template):
```
Comment (fblanchetNaN) [inline incipyt/project/structure.py:105]: Why do you prefer classmethod over staticmethod here?

File: incipyt/project/structure.py [hunk 17]
```diff
@@ -1,147 +1,13 @@
...
-    def __ior__(self, other):
-        self.update(other)
-        return self
-
-
-environ = _Environ()
-
-
-class _Structure:
+class Structure:
```
Comment (fblanchetNaN) [inline incipyt/project/structure.py:10]: Why remove leading _ while you import it later with `as _` ?

Comment (Julien00859) [reply incipyt/project/structure.py:10]: I don't like private classes (hence no `_`) but I didn't want to break compatibility (hence the import as)

Comment (Julien00859) [reply incipyt/project/structure.py:10]: oh but that's singleton actually :facepalm: I should had kept the private class
